### ISSUE TITLE
[Question/Feature Request] Download entire requirements tree with 'composer archive'

### ISSUE THREAD

**Issue by mooror** [composer/composer#6119]
Hello there everyone. I work at a company that uses a private satis repository to store our private composer packages. Occasionally we require the use of a package from packagist.org that has many requirements. In these situations it makes no sense to manually download and store the package and all of its requirements. Therefor we have been looking for an alternative that will download the package and its requirements as zip files and store them in our private repository. Our first thought was to use satis's built in mirroring functionality. We did test runs using the archive array (satis's downloading functionality)  in concert with the `"require-dependencies": true` key/pair. But we found (due to misconfiguration or design) that the mirroring functionality parsed **all** the requirements that matched the version string of the requirements. So when the target package had many requirements like this:
```
    "user/package": "~2.7",
``` 
satis seemed to parse all the versions that matched the version string instead of just grabbing the latest version of "user/package" that matched it. This as you can imagine caused problems because of the sheer amount of packages being parsed. So we went in search of another solution and discovered `composer archive` which brings us to the present moment. 

**Primary Question:** We are wondering if it would be possible to add a feature to the archive command (or another composer command) that would allow developers to not only download the package but also its entire requirement tree (in separate zip files of course) in one call. Something like `composer archive user/package ~2.7 --format=zip --include-requirements`. I'm not familiar with how composer works internally but I imagine that the composer install command downloads packages as zip or tar files and then unzips them to there proper location. If this is indeed the case then (in theory) this functionality would not be to hard to create as you could use all the current composer functionality and just forgo unzipping once the packages are on the local machine. 

I would be willing to have a go at creating this functionality but I would need some help as I am not familiar with composers internals.

**Secondary Question:** While we do believe that building this functionality directly into composer would be useful, we also understand that there are many other features and functionalities that are in the works so this may not be of a high priority. Therefor we have thought of creating this functionality through a python or php script (if necessary). My second question then, is how would multiple calls to composers archive command (in a loop or via child processes) effect performance when compared with a single call like 'composer install'.

**Comment by alcohol** [comment]
Satis just has to download those other versions once. What is the big problem?

**Comment by mooror** [comment]
Hello there alcohol, 
First, let me just thank you for all the hard work you have put into satis. You have saved my company's developer team a lot of time and for that I am very grateful. 

Now as for your question, I am not 100% sure that we are on the same page. So before I answer your question let me explain exactly what we are trying to achieve and how we tried to do that using satis. 

The company I work for uses a private satis repository to keep proprietary packages away from the public. Specifically we use an archive repository using the following (in satis.json):
```
{
    "name": "Private Repository",
    "homepage": "http://repository.oursite.ca",
    "repositories": [
 {
            "type": "artifact",
            "url": "/folder/path/to/artifact-repository/"
        }

    ],
    "require-all": true
}
```
This works for use most of the time but sometimes we needed public packages (from packagist) to aid us in our development. So (up until now) we would manually download these packagist packages from their github repositories. This (as you can imagine) was very time consuming, but since most of the packages we needed had minimal requirements it was doable. But when we recently needed to use a framework package with many requirements (each having there own requirements) we realized just how deep the rabbit hole goes and starting looking for an automated solution. This lead to our discovery of satis's mirroring functionality which lead to the test code below (in satis.json):
```
{
    "name": "Private Repository",
    "homepage": "http://repository.oursite.ca",
"repositories": [
 {
      "type": "composer",
      "url": "https://packagist.org"
    }
  ],
  "archive": {
    "directory": "dist",
    "format": "zip",
    "skip-dev": true
  },
"require": {
    "silverstripe/framework": "4.0.0-alpha4"
  },
"require-dependencies": true

}
```
In this test run we where trying to get the silverstripe framework package downloaded with all of its requirements. But we only want the latest version (that matches the version constraint) of each of those requirements. Heres the silverstripe framework requirements:
```
"require": {
    "php": ">=5.5.0",
    "composer/installers": "~1.0",
    "monolog/monolog": "~1.11",
    "league/flysystem": "~1.0.12",
    "symfony/yaml": "~2.7",
    "embed/embed": "^2.6",
    "swiftmailer/swiftmailer": "~5.4",
    "symfony/config": "^2.8",
    "symfony/translation": "^2.8",
    "vlucas/phpdotenv": "^2.4"
},
```
And heres the output of the satis build (using the test config.yaml file)
[Output (Gist)](https://gist.github.com/mooror/395a09e459ed74f246b528b86ee13f56)
Its quite long, but if you do a search for 'Installing monolog/monolog' (using ctrl + f) you will find that it appears 6 different times. This may make sense logically (seeing that the version constraint is "~1.11") and may even fit the needs of many satis users. But from my standpoint, downloading 6 different versions of a single package when you only need the latest for that constraint is inefficient. I don't mean to be rude in any way, and like I said previously, this may work for many satis users. But for us, this solution would eat up too much storage space, especially as we start to use more frameworks and toolkits.

So hopefully this answers your question and helps you to see where we are coming from and what are goals are. 

P.S. Sorry if this seems a tad repetitive. I just wanted to be super clear about our situation.

**Comment by mooror** [comment]
@alcohol Also we would also be open to any ideas you might have on how we might accomplish our goals using satis. In fact I believe it might be more appropriate to do this kind of thing in satis. I had just thought that this might be easier to implement in composer.

**Comment by alcohol** [comment]
I understand what you want. Unfortunately, this is simply not the way Satis is designed to work.

If you want to limit the versions that Satis downloads to only the latest version, you will have to setup a specific version constraint in the `require` directory for each and every package that the framework depends on (and all sub-dependencies). You can determine which versions that would be by simply running `composer show` in your project directory after you installed all dependencies. But that would still require a rather long and tedious process of adding all these package names + specific version requirements into your satis.json configuration file.

Also, you talk about the many versions consuming disk storage, but honestly, with a terrabyte disk costing next to nothing now a days, I do not quite understand why disk storage should be of any concern at all.

TL/DR: what you are asking for is currently not supported, and I don't see anyone adding this functionality in the near future either. Satis development is almost entirely user-driven now a days. We (@composer team) merely merge pull requests. You are more than welcome to invest time into developing this feature yourself and submitting a pull request. I will happily merge it for you (provided the changes introduced are within reason).

PS. This issue is probably more suited for the [`composer/satis`](/composer/satis) repository.

**Comment by mooror** [comment]
> PS. This issue is probably more suited for the composer/satis repository.

Like I said in the above comment, I too believe this functionality would be more appropriate for satis. I had also guessed that this would be mostly up to myself to create. Therefor, though it would benefit satis, I believe I could achieve the functionality much quicker if I base it off of composer's install functionality. But in order to confirm whether or not that is the case, I need some information. 

**Question 1:** How does the composer archive command work? 

**Question 2:** Does the composer `install command` transfer download packages in .zip format?

**Question 3:** When you run composer install how does the command determine the latest version that satisfies the version restraint?

I will be researching into these questions myself, but any help in that area would be great. I will post my discoveries here so you can confirm or deny my findings. I don't want to start developing something without having a decent understanding of how the system works.

P.S. Due to time constraints I will attempt to develop this functionality in the quickest way possible. Currently I believe that will mean I base it on composer functionality, but if it looks to be quicker implementing it on satis after I do some more research, then so be it.

**Comment by mooror** [comment]
> Question 3: When you run composer install how does the command determine the latest version that satisfies the version constraint?

From the official docs:
_If there is a composer.lock file in the current directory, it will use the exact versions from there instead of resolving them. [...]
If there is no composer.lock file, Composer will create one after dependency resolution._

While I'm not sure yet as to how 'dependency resolution' works exactly, the docs seem to indicate that this is done through the `composer update` command.

Update command docs:
_In order to get the latest versions of the dependencies and to update the composer.lock file, you should use the update command.[...]
This will resolve all dependencies of the project and write the exact versions into composer.lock._

**Comment by alcohol** [comment]
Honestly, simply adding the specific versions you need to your satis.json will be much faster if you look at hours invested, vs researching and building a solution for a problem that is barely a problem at all.
